### PR TITLE
Bring parity with pandas in Index.join

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
This PR removes materialization of certain `RangeIndex.join` combinations by falling back to pandas because it's not expensive to do so and also fixes an issue around mismatching index names while performing a join.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/index.py [hunk 1]
```diff
@@ -839,6 +839,27 @@ def join(
         # TODO: pandas supports directly merging RangeIndex objects and can
         # intelligently create RangeIndex outputs depending on the type of
         # join. We need to implement that for the supported special cases.
```
Comment (shwina) [inline python/cudf/cudf/core/index.py:841]: Should we remove this TODO now?

Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:841]: Yes good point, will remove this todo.

**Review by galipremsagar — COMMENTED** [review]
Need to add pytests, WIP

File: python/cudf/cudf/core/index.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py]: Removed this todo

File: python/cudf/cudf/core/index.py [hunk 2]
```diff
@@ -836,9 +836,32 @@ def _binaryop(self, other, op: str):
...
+        if how in {"left", "right"} or self.equals(other):
+            # pandas supports directly merging RangeIndex objects and can
+            # intelligently create RangeIndex outputs depending on the type of
+            # join. Hence falling back to performing a merge on pd.RangeIndex
+            # since the conversion is cheap.
+            self_pd = self.to_pandas()
+            if isinstance(other, RangeIndex):
+                other_pd = other.to_pandas()
+            elif isinstance(other, pd.RangeIndex):
+                other_pd = other
```
Comment (galipremsagar) [inline python/cudf/cudf/core/index.py:848]:
```suggestion
```

File: python/cudf/cudf/core/index.py [hunk 3]
```diff
@@ -836,9 +836,30 @@ def _binaryop(self, other, op: str):
...
+            # pandas supports directly merging RangeIndex objects and can
+            # intelligently create RangeIndex outputs depending on the type of
+            # join. Hence falling back to performing a merge on pd.RangeIndex
+            # since the conversion is cheap.
+            self_pd = self.to_pandas()
+            if isinstance(other, RangeIndex):
+                other_pd = other.to_pandas()
+            else:
+                other_pd = None
+            if other_pd is not None:
```
Comment (shwina) [inline python/cudf/cudf/core/index.py:849]: I believe we can just do this?

```suggestion
```

Comment (shwina) [inline python/cudf/cudf/core/index.py:849]: I believe we can just do this?

```suggestion
```

File: python/cudf/cudf/core/index.py [hunk 4]
```diff
@@ -836,9 +836,27 @@ def _binaryop(self, other, op: str):
...
+        if how in {"left", "right"} or self.equals(other):
+            # pandas supports directly merging RangeIndex objects and can
+            # intelligently create RangeIndex outputs depending on the type of
+            # join. Hence falling back to performing a merge on pd.RangeIndex
+            # since the conversion is cheap.
+            self_pd = self.to_pandas()
+            if isinstance(other, RangeIndex):
+                other_pd = other.to_pandas()
+                result = self_pd.join(
+                    other_pd,
```
Comment (galipremsagar) [inline python/cudf/cudf/core/index.py:848]:
```suggestion
                result = self_pd.join(
                    other.to_pandas(),
```

File: python/cudf/cudf/core/index.py [hunk 5]
```diff
@@ -836,9 +836,26 @@ def _binaryop(self, other, op: str):
...
-        # intelligently create RangeIndex outputs depending on the type of
-        # join. We need to implement that for the supported special cases.
+        if how in {"left", "right"} or self.equals(other):
+            # pandas supports directly merging RangeIndex objects and can
+            # intelligently create RangeIndex outputs depending on the type of
+            # join. Hence falling back to performing a merge on pd.RangeIndex
+            # since the conversion is cheap.
+            self_pd = self.to_pandas()
+            if isinstance(other, RangeIndex):
+                result = self_pd.join(
```
Comment (galipremsagar) [inline python/cudf/cudf/core/index.py:846]:
```suggestion
            if isinstance(other, RangeIndex):
                result = self.to_pandas().join(
```

File: python/cudf/cudf/core/index.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py]: Yup, removed it

**Review by shwina — APPROVED** [review]

**Pull request merged**